### DISCUSSION TITLE
History computation issue

### DISCUSSION THREAD

**Discussion by fbiville** [jreleaser/jreleaser#956] [Q&A] [answered]
Hello, 

I'm not sure what is wrong, but the generated change log for https://github.com/liquibase/liquibase-neo4j includes too many commits.

# The Setup

Here is the generated change log (note: [the actual one](https://github.com/liquibase/liquibase-neo4j/releases/tag/liquibase-neo4j-4.16.1.1) has been manually fixed up) for version 1.16.1.1:

```markdown
## Changelog

- 12c3da7 Set next release version (Florent Biville)
- 41fbfa1 Fix contribution exclusion from change log (Florent Biville)
- d8cfdef Support <loadData /> (Andrea Santurbano)
- e60822b Version Bumped to Snapshot for Development (liquibot)
- faf38ae Version Bumped to 4.16.1 (liquibot)
- 745ed16 Bump liquibase-core from 4.16.0 to 4.16.1 (dependabot[bot])
- c60e0b0 Bump maven-shade-plugin from 3.3.0 to 3.4.0 (dependabot[bot])
- 6eb9ed3 Bump slf4j-nop from 1.7.36 to 2.0.1 (dependabot[bot])
- c6ffe90 Add missing copyright (Florent Biville)
- 1a4dd9f Version Bumped to Snapshot for Development (liquibot)
- 62bcf1a Version Bumped to 4.16.0 (liquibot)
- f3d561e Bump liquibase-core from 4.15.0 to 4.16.0 (dependabot[bot])


## Contributors
We'd like to thank the following people for their contributions:
dependabot[bot], Andrea Santurbano, Florent Biville, liquibot
```

The Maven plugin is [configured](https://github.com/liquibase/liquibase-neo4j/blob/5b6148e1e8e8ed47789a73815ba64edf1c8f3442/pom.xml#L340-L376) as follows:

```xml
<release>
                                    <github>
                                        <!-- [...] -->
                                        <changelog>
                                            <preset>conventional-commits</preset>
                                            <formatted>ALWAYS</formatted>
                                            <hide>
                                                <contributors>
                                                    <contributor>GitHub</contributor>
                                                    <contributor>.*bot</contributor>
                                                </contributors>
                                            </hide>
                                        </changelog>
                                    </github>
                                </release>
```

The plugin is followed when the `published` event is triggered in GitHub Actions (full workflow [here](https://github.com/liquibase/liquibase-neo4j/blob/5b6148e1e8e8ed47789a73815ba64edf1c8f3442/.github/workflows/release-published.yml)):

```yaml
        - name: Populate release change log
          run: mvn --quiet --batch-mode jreleaser:release -P release
          env:
            JRELEASER_GITHUB_TOKEN: ${{ secrets.JRELEASER_GITHUB_TOKEN }}
```

# History computation issue

The changelog above was generated for 1.6.1.1 (i.e. for tag ` liquibase-neo4j-4.16.1.1`).
This version follows 1.6.1, i.e. `liquibase-neo4j-4.16.1` tag.
When manually running the diff, here is what I get:

```shell
$ git log liquibase-neo4j-4.16.1..liquibase-neo4j-4.16.1.1 --oneline
12c3da7 (tag: liquibase-neo4j-4.16.1.1) Set next release version
41fbfa1 Fix contribution exclusion from change log
d8cfdef Support <loadData />
e60822b Version Bumped to Snapshot for Development
faf38ae Version Bumped to 4.16.1
745ed16 Bump liquibase-core from 4.16.0 to 4.16.1
c60e0b0 Bump maven-shade-plugin from 3.3.0 to 3.4.0
6eb9ed3 Bump slf4j-nop from 1.7.36 to 2.0.1
```

The generated change log includes these, but also 4 previous commits for reasons I do not understand:

```shell
- c6ffe90 Add missing copyright (Florent Biville)
- 1a4dd9f Version Bumped to Snapshot for Development (liquibot)
- 62bcf1a Version Bumped to 4.16.0 (liquibot)
- f3d561e Bump liquibase-core from 4.15.0 to 4.16.0 (dependabot[bot])
```

These are from previous releases.

**Comment by fbiville** [comment, marked as answer]
Actually, there is something fishy going on here wrt the repo history...
I suspect some tags (automatically created by liquibot) are on merge commits, and since I rebase and merge, these merge commits are lost :|

For instance, 4.16.1 is commit `7aa7d29`, but this commit is not available on any branch.
It is however part of https://github.com/liquibase/liquibase-neo4j/pull/235, which was rebased before merging ("Rebase and merge"), so the commit disappears and JReleaser cannot compute the diff properly.

**Reply by fbiville** [reply]
yet another rubber duck debugging discussion 😭

**Reply by aalmiray** [reply]
🦆

**Comment by aalmiray** [comment]
Please have a look at `target/jreleaser/trace.log` as it contains info used by the changelog generator such as the auto-calculated `previousTag`. This will tell you if the previous tag was the one you expect or if it was a different one. FWIW you can explicitly configure he value for `previousTag`.
